File: android-java-foundation/src/main/java/com/frontend/notify/Notifier.java
// ======================================================================
// Project Name    : android_foundation
//
// Copyright © 2018 U-CREATES. All rights reserved.
//
// This source code is the property of U-CREATES.
// If such findings are accepted at any time.
// We hope the tips and helpful in developing.
// ======================================================================
package com.frontend.notify;
import com.core.entity.component.Parameter;
import java.util.HashMap;
public class Notifier {
    private static Notifier instance = null;
    private HashMap<Integer, INotifier> notifierHashMap = new HashMap<Integer, INotifier>();
    public static Notifier getInstance() {
        if (null == Notifier.instance) {
            Notifier.instance = new Notifier();
        }
        return Notifier.instance;
    }
    public void notify(NotifyMessage message) {
        this.notify(message, null);
        return;
    }
    public void notify(NotifyMessage message, Parameter parameter) {
        for (Integer id : this.notifierHashMap.keySet()) {
            INotifier inotifier = this.notifierHashMap.get(id);
            inotifier.onNotify(message, parameter);
        }
        return;
    }
    public boolean add(INotifier inotifier) {
        int uniqId = System.identityHashCode(inotifier);
        if (false == this.notifierHashMap.containsKey(uniqId)) {
            this.notifierHashMap.put(uniqId, inotifier);
            return true;
        }
        return false;
    }
}

File: android-kotlin-foundation/src/main/java/com/frontend/notify/Notifier.kt
// ======================================================================
// Project Name    : android_foundation
//
// Copyright © 2018 U-CREATES. All rights reserved.
//
// This source code is the property of U-CREATES.
// If such findings are accepted at any time.
// We hope the tips and helpful in developing.
// ======================================================================
package com.frontend.notify
import com.core.entity.component.Parameter
open class Notifier {
    companion object {
        private val instance: Notifier = Notifier()
        private var notifierHashMap: HashMap<Int, INotify> = HashMap<Int, INotify>()
        fun getInstance(): Notifier {
            return this.instance
        }
        fun notify(message: NotifyMessage) {
            this.notify(message, null)
            return
        }
        fun notify(message: NotifyMessage, parameter: Parameter?) {
            this.notifierHashMap.forEach { (key, inotify) ->
                inotify.onNotify(message, parameter)
            }
            return
        }
        fun add(notifier: INotify): Boolean {
            val uniqId: Int = System.identityHashCode(notifier)
            if (false == this.notifierHashMap.containsKey(uniqId)) {
                this.notifierHashMap.put(uniqId, notifier)
                return true
            }
            return false
        }
    }
}